### PR TITLE
Unfold->conv for morphology backbone

### PR DESCRIPTION
**Pull request by ducha-aiki**
### Description
Changes the morphology backbone for less memory consumption
Related: https://github.com/kornia/kornia/pull/1034#issuecomment-861525168 

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

### CODE REVIEW

File: kornia/morphology/morphology.py [hunk 1]
```diff
@@ -151,10 +161,13 @@ def erosion(
...
 
-    output = output.unfold(2, se_h, 1).unfold(3, se_w, 1)
-    output, _ = torch.min(output - neighborhood, 4)
-    output, _ = torch.min(output, 4)
-
+    B, C, H, W = tensor.size()
+    Hpad, Wpad = output.shape[-2:]
+    reshape_kernel = _neight2channels_like_kernel(kernel)
+    output, _ = F.conv2d(
+        output.view(B * C, 1, Hpad, Wpad), reshape_kernel, padding=0, bias=-neighborhood.view(-1).flip(0)
```
Comment (Manza12) [inline kornia/morphology/morphology.py:168]: erosion do not need to flip structuring element

**Review by Manza12 — CHANGES_REQUESTED** [review]
Erosion do not need to flip the structuring element

File: kornia/morphology/morphology.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/morphology/morphology.py:168]: Hmmm, but the tests were not failed because of my flip. Could you please add one, which covers this?

Comment (Manza12) [reply kornia/morphology/morphology.py]: I am not able to PR my tests :(

Comment (Manza12) [reply kornia/morphology/morphology.py]: This is the new test for dilation:
```
    def test_flip(self, device, dtype):
        tensor = torch.tensor([[0.5, 1.0, 0.3], [0.7, 0.3, 0.8], [0.4, 0.9, 0.2]], device=device, dtype=dtype)[
            None, None, :, :
        ]
        kernel = torch.tensor([[0.0, 1.0, 1.0], [0.0, 1.0, 1.0], [0.0, 1.0, 1.0]], device=device, dtype=dtype)
        expected = torch.tensor([[0.7, 1.0, 1.0], [0.7, 1.0, 1.0], [0.7, 0.9, 0.9]], device=device, dtype=dtype)[
            None, None, :, :
        ]
        assert_allclose(dilation(tensor, kernel), expected)
```

This is the new test for erosion:
```
        def test_flip(self, device, dtype):
        tensor = torch.tensor([[0.5, 1.0, 0.3], [0.7, 0.3, 0.8], [0.4, 0.9, 0.2]], device=device, dtype=dtype)[
            None, None, :, :
        ]
        kernel = torch.tensor([[0.0, 1.0, 1.0], [0.0, 1.0, 1.0], [0.0, 1.0, 1.0]], device=device, dtype=dtype)
        expected = torch.tensor([[0.3, 0.3, 0.3], [0.3, 0.2, 0.2], [0.3, 0.2, 0.2]], device=device, dtype=dtype)[
            None, None, :, :
        ]
        assert_allclose(dilation(tensor, kernel), expected)
```

Comment (ducha-aiki) [reply kornia/morphology/morphology.py]: Thank you!

**Review by Manza12 — APPROVED** [review]
Great job!

File: docs/generate_example_images.py [hunk 2]
```diff
@@ -11,7 +11,8 @@
 mod = importlib.import_module("kornia.augmentation")
 
 BASE_IMAGE_URL: str = "https://raw.githubusercontent.com/kornia/data/main/panda.jpg"
-OUTPUT_PATH = Path(__file__).absolute().parent / "source/_static"
+OUTPUT_PATH = Path(__file__).absolute().parent / "build/html/_static"
```
Comment (shijianjian) [inline docs/generate_example_images.py:14]: This should be irrelevant to this PR.

File: kornia/morphology/morphology.py [hunk 3]
```diff
@@ -74,9 +80,13 @@ def dilation(
         neighborhood = structuring_element.clone()
```
Comment (shijianjian) [inline kornia/morphology/morphology.py:74]: Why do we need to clone? Potentially be not updatable whilst optimization?

Comment (ducha-aiki) [reply kornia/morphology/morphology.py:74]: Otherwise it is not differentiable

File: docs/generate_example_images.py (hunk 2 above)
Comment (ducha-aiki) [reply docs/generate_example_images.py:14]: Good catch, no idea, how it ended up here...

**Comment by Manza12** [comment]
I did clone because I didn’t know how to do it in other way; the point is the next:

  *   Your structuring element is a tensor with all his values to floating (potentially updatable) points
  *   But the kernel says which points do actually matter, so what it does is to put unused points to minus infinity such that they are omitted by minimum/maximum
In order to optimize that, the better way is to put an if statement in the kernel that performs the minimum/maximum operation to avoid taking into account unused values.

I didn’t know how to do it in other fashion in PyTorch.
Maybe just using neighbourhood=structuring element. The single problem is that then you will change values of the structuring element (putting them to minus infinity) so I don’t know if this is suitable...

Is there any workaround ?

De: Jian ***@***.***>
Enviado: viernes, 18 de junio de 2021 15:11
Para: ***@***.***>
CC: ***@***.***>; Review ***@***.***>
Asunto: Re: [kornia/kornia] Unfold->conv for morphology backbone (#1107)


@shijianjian commented on this pull request.


In docs/generate_example_images.py<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkornia%2Fkornia%2Fpull%2F1107%23discussion_r654413084&data=04%7C01%7C%7C31a465abbf73489ab13508d9325aa1c4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637596187103822136%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=AK9sIpDRprc1RM4YPVdRfDz%2FXGOWlb%2B7K%2FIzazENdVE%3D&reserved=0>:

> @@ -11,7 +11,8 @@

 mod = importlib.import_module("kornia.augmentation")



 BASE_IMAGE_URL: str = "https://raw.githubusercontent.com/kornia/data/main/panda.jpg"

-OUTPUT_PATH = Path(__file__).absolute().parent / "source/_static"

+OUTPUT_PATH = Path(__file__).absolute().parent / "build/html/_static"

This should be irrelevant to this PR.


In kornia/morphology/morphology.py<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkornia%2Fkornia%2Fpull%2F1107%23discussion_r654415711&data=04%7C01%7C%7C31a465abbf73489ab13508d9325aa1c4%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637596187103822136%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=%2FaHVDUFMUOH2I1ZIuXfTLjntwqhsg2WIWB6DZMDVKeY%3D&reserved=0>:

> @@ -74,9 +80,13 @@ def dilation(

         neighborhood = structuring_element.clone()

Why do we need to clone? Potentially be not updatable whilst optimization?

**Comment by shijianjian** [comment]
> The single problem is that then you will change values of the structuring element (putting them to minus infinity) so I don’t know if this is suitable... Is there any workaround ?

It is alright if you are doing it on purpose. I was only concerning if it will lose track of those value changes.

**Review by shijianjian — APPROVED** [review]

File: kornia/morphology/morphology.py [hunk 4]
```diff
@@ -74,9 +80,13 @@ def dilation(
...
         neighborhood[kernel == 0] = -max_val
 
-    output = output.unfold(2, se_h, 1).unfold(3, se_w, 1)
-    output, _ = torch.max(output + neighborhood.flip((0, 1)), 4)
-    output, _ = torch.max(output, 4)
+    B, C, H, W = tensor.size()
+    Hpad, Wpad = output.shape[-2:]
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:84]:
```suggestion
    h_pad, w_pad = output.shape[-2:]
```

File: kornia/morphology/morphology.py (hunk 3 above)
Comment (edgarriba) [reply kornia/morphology/morphology.py:74]: why ? this is a weird one

File: kornia/morphology/morphology.py [hunk 5]
```diff
@@ -74,9 +80,13 @@ def dilation(
...
+    reshape_kernel = _neight2channels_like_kernel(kernel)
+    output, _ = F.conv2d(
+        output.view(B * C, 1, Hpad, Wpad), reshape_kernel, padding=0, bias=neighborhood.view(-1).flip(0)
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:87]:
```suggestion
        output.view(B * C, 1, h_pad, w_pad), reshape_kernel, padding=0, bias=neighborhood.view(-1).flip(0)
```

File: kornia/morphology/morphology.py [hunk 6]
```diff
@@ -74,9 +80,13 @@ def dilation(
...
+    ).max(dim=1)
+    output = output.view(B, C, H, W)
 
     return output
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:91]:
```suggestion
    return output.view_as(tensor)
```

File: kornia/morphology/morphology.py [hunk 7]
```diff
@@ -4,6 +4,12 @@
 import torch.nn.functional as F
 
 
+def _neight2channels_like_kernel(kernel):
```
Comment (edgarriba) [inline kornia/morphology/morphology.py:7]:
```suggestion
def _neight2channels_like_kernel(kernel: torch.Tensor) -> torch.Tensor:
```

**Comment by edgarriba** [comment]
@ducha-aiki could you rebase ?

File: kornia/morphology/morphology.py (hunk 5 above)
Comment (ducha-aiki) [reply kornia/morphology/morphology.py:87]: clone - because we are replacing the values in the next line. Without such thing, it fails if we try to backprop

File: kornia/morphology/morphology.py (hunk 3 above)
Comment (ducha-aiki) [reply kornia/morphology/morphology.py:74]: Because below we are in-lace replacing the values

**Comment by edgarriba** [comment]
@ducha-aiki @Manza12 are the precision issues expected with the convolutional approach?

**Comment by ducha-aiki** [comment]
> @ducha-aiki @Manza12 are the precision issues expected with the convolutional approach?

Unfortunately yes, esp on cuda - it is price for less memory consumption. If you think it is more important, we might not merge this

**Comment by edgarriba** [comment]
@ducha-aiki can we have a flag `method` with an enum `MorphologyKeys.FAST` and `MorphologyKeys.ACCURATE` or similar ? @shijianjian @lferraz @Manza12

**Comment by Manza12** [comment]
I don't know about this precision topics. Why CUDA does that?

________________________________

**Comment by shijianjian** [comment]
> @ducha-aiki can we have a flag `method` with an enum `MorphologyKeys.FAST` and `MorphologyKeys.ACCURATE` or similar ? @shijianjian @lferraz @Manza12

It could be a good approach. Shall we introduce something similar to opencv like ```K.MORPH_FAST```? or as an independent class?

**Comment by edgarriba** [comment]
@shijianjian @Manza12 @ducha-aiki let's investigate for this precision issue. If it's clearly a problem of the convolutional approach, then we can introduce dual implementation specified by enum flags.

**Comment by Manza12** [comment]
Can we meet sometime? I don't exactly know what is going on
________________________________

**Comment by edgarriba** [comment]
@ducha-aiki @Manza12 can we unblock this ?

**Comment by ducha-aiki** [comment]
It is definitely convolution precision approach. What do you propose for the enum flag, to let user decide the algorithm?

**Comment by edgarriba** [comment]
enum seems fine

**Comment by ducha-aiki** [comment]
I actually don't see why enum and not string

**Comment by edgarriba** [comment]
> I actually don't see why enum and not string

more safe but could work too
